File: frontend/src/utils/exportCsv.js
/** Build and download a CSV file in the browser. */
export const downloadCsv = (filename, headers, rows) => {
  const escape = (value) => {
    const text = value == null ? '' : String(value);
    if (/[",\n]/.test(text)) {
      return `"${text.replace(/"/g, '""')}"`;
    }
    return text;
  };

  const lines = [
    headers.map(escape).join(','),
    ...rows.map((row) => row.map(escape).join(',')),
  ];
  const blob = new Blob([lines.join('\n')], { type: 'text/csv;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  link.click();
  URL.revokeObjectURL(url);
};

export const exportOrdersCsv = (data) => {
  if (!data?.orders?.length) return;
  downloadCsv(
    'orders.csv',
    ['Order', 'Customer', 'Mobile', 'Date', 'Status', 'Payment', 'Revenue', 'Profit', 'Items'],
    data.orders.map((order) => [
      order.order_number,
      order.customer?.name,
      order.customer?.mobile,
      order.order_date,
      order.status,
      order.payment_status,
      order.total_revenue,
      order.total_profit,
      (order.matching_items || [])
        .map((i) => `${i.product_name} ${i.product_unit} x${i.quantity}`)
        .join('; '),
    ]),
  );
};

export const exportCustomersCsv = (data) => {
  if (!data?.customers?.length) return;
  downloadCsv(
    'customers.csv',
    ['Rank', 'Customer', 'Mobile', 'Orders', 'Revenue', 'Profit'],
    data.customers.map((c, i) => [
      i + 1,
      c.customer_name,
      c.mobile,
      c.order_count,
      c.total_spent,
      c.total_profit,
    ]),
  );
};

export const exportProductsCsv = (data) => {
  if (!data?.products?.length) return;
  downloadCsv(
    'products.csv',
    ['Product', 'Unit', 'Category', 'Price', 'Margin %'],
    data.products.map((p) => [p.name, p.unit, p.category, p.selling_price, p.margin_percent]),
  );
};

export const exportOfferingsCsv = (data) => {
  if (!data?.offerings?.length) return;
  const rows = [];
  data.offerings.forEach((offering) => {
    (offering.items || []).forEach((item) => {
      rows.push([
        offering.offering_date,
        offering.status,
        item.product_name,
        item.product_unit,
        item.category,
        item.selling_price,
        item.available_quantity ?? '',
      ]);
    });
  });
  downloadCsv(
    'daily-offerings.csv',
    ['Date', 'Status', 'Product', 'Unit', 'Category', 'Price', 'Available Qty'],
    rows,
  );
};

export const exportPaymentTrendsCsv = (data) => {
  if (!data?.by_method) return;
  const total = parseFloat(data.total_amount || 0);
  downloadCsv(
    'payment-trends.csv',
    ['Method', 'Count', 'Amount', 'Share %'],
    Object.entries(data.by_method).map(([method, stats]) => {
      const amount = parseFloat(stats.amount || 0);
      const share = total ? ((amount / total) * 100).toFixed(1) : '0';
      return [stats.label || method, stats.count, amount, share];
    }),
  );
};

export const exportResultCsv = (type, data) => {
  switch (type) {
    case 'orders':
      exportOrdersCsv(data);
      break;
    case 'customers':
      exportCustomersCsv(data);
      break;
    case 'products':
      exportProductsCsv(data);
      break;
    case 'offerings':
      exportOfferingsCsv(data);
      break;
    case 'payment_trends':
      exportPaymentTrendsCsv(data);
      break;
    default:
      break;
  }
};

File: assistant/export_csv.py
"""Export assistant result data to CSV."""

from __future__ import annotations

import csv
import io
from typing import Any


def _rows_for_type(response_type: str, data: dict[str, Any]) -> tuple[list[str], list[list[Any]]]:
    if response_type == 'orders' and data.get('orders'):
        headers = [
            'order_number', 'customer_name', 'mobile', 'order_date', 'status',
            'payment_status', 'revenue', 'profit', 'items',
        ]
        rows = []
        for order in data['orders']:
            items = '; '.join(
                f"{i.get('product_name')} {i.get('product_unit')} x{i.get('quantity')}"
                for i in (order.get('matching_items') or [])
            )
            rows.append([
                order.get('order_number'),
                order.get('customer', {}).get('name'),
                order.get('customer', {}).get('mobile'),
                order.get('order_date'),
                order.get('status'),
                order.get('payment_status'),
                order.get('total_revenue'),
                order.get('total_profit'),
                items,
            ])
        return headers, rows

    if response_type == 'customers' and data.get('customers'):
        headers = [
            'rank', 'customer_name', 'mobile', 'orders', 'revenue', 'profit',
            'apartment', 'block',
        ]
        rows = []
        for i, c in enumerate(data['customers'], start=1):
            rows.append([
                i,
                c.get('customer_name'),
                c.get('mobile'),
                c.get('order_count'),
                c.get('total_spent'),
                c.get('total_profit'),
                c.get('apartment_name'),
                c.get('block'),
            ])
        return headers, rows

    if response_type == 'products' and data.get('products'):
        headers = ['name', 'unit', 'category', 'price', 'margin_percent']
        rows = [
            [p.get('name'), p.get('unit'), p.get('category'), p.get('selling_price'), p.get('margin_percent')]
            for p in data['products']
        ]
        return headers, rows

    if response_type == 'offerings' and data.get('offerings'):
        headers = ['offering_date', 'status', 'product', 'unit', 'category', 'price', 'available_qty']
        rows = []
        for offering in data['offerings']:
            for item in offering.get('items') or []:
                rows.append([
                    offering.get('offering_date'),
                    offering.get('status'),
                    item.get('product_name'),
                    item.get('product_unit'),
                    item.get('category'),
                    item.get('selling_price'),
                    item.get('available_quantity'),
                ])
        return headers, rows

    if response_type == 'payment_trends' and data.get('by_method'):
        headers = ['method', 'count', 'amount', 'share_percent']
        total = float(data.get('total_amount') or 0)
        rows = []
        for method, stats in data['by_method'].items():
            amount = float(stats.get('amount') or 0)
            share = round(amount / total * 100, 1) if total else 0
            rows.append([method, stats.get('count'), amount, share])
        return headers, rows

    return [], []


def result_to_csv(response_type: str, data: dict[str, Any] | None) -> str:
    headers, rows = _rows_for_type(response_type, data or {})
    output = io.StringIO()
    writer = csv.writer(output)
    if headers:
        writer.writerow(headers)
        writer.writerows(rows)
    return output.getvalue()
